File: Crio/crio_programming_interview_problems-master/StackTheBoxes/StackTheBoxes.java
import java.util.*; 
//CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
class Triplet {
    int first,second,third;
    void initTriplet(int first,int second, int third) {
        this.first = first;
        this.second = second;
        this.third = third;
    }
}
class TripletComparator implements Comparator<Triplet> { 
    public int compare(Triplet o1, Triplet o2) { 
          if(o1.first < o2.first) {
             return -1;
          } else {
            if(o1.first == o2.first) {
                if(o1.second < o2.second) {
                    return -1;
                } else {
                    if(o1.second == o2.second) {
                        if(o1.third < o2.third) {
                            return -1;
                        } else {
                            return 1;
                        }
                    } else {
                        return 1;
                    }
                }
            } else {
                return 1;
            }
          } 
    }  
}
//CRIO_SOLUTION_END_MODULE_L1_PROBLEMS
public class StackTheBoxes {
    //CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
    int findLisSum( ArrayList<Triplet> triplets , int n ) {
        int ans[] = new int[n];
        for(int i = n-1 ; i>= 0 ; i--) {
            ans[i] = 0;
            for(int j = i+1 ; j < n ; j++) {
                if(triplets.get(i).first < triplets.get(j).first && triplets.get(i).second < triplets.get(j).second && triplets.get(i).third < triplets.get(j).third) {
                    ans[i] = Math.max(ans[i], ans[j]);
                }
            }
            ans[i] += triplets.get(i).first;
        }
        int maxHeight = 0;
        for(int i = 0 ; i < n ;i++) {
            maxHeight = Math.max(maxHeight,ans[i]);
        }
        return maxHeight;
    }
    //CRIO_SOLUTION_END_MODULE_L1_PROBLEMS
    int stackTheBoxes(int length[] , int breadth[] , int height[],int n) {
        //CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
        ArrayList<Triplet> triplets = new ArrayList<Triplet>();
        for(int i = 0 ; i < n ; i++) {
            Triplet temp = new Triplet();
            temp.initTriplet(height[i], breadth[i], length[i]);
            triplets.add(temp);
        }
        Collections.sort(triplets,new TripletComparator());
        int maxHeight = findLisSum(triplets, n);
        return maxHeight;  
        //CRIO_SOLUTION_END_MODULE_L1_PROBLEMS
    }
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        int length[] = new int[n];
        int breadth[] = new int[n];
        int height[] = new int[n];
        for(int i = 0 ; i < n ;i++) {
            length[i] = scanner.nextInt();
            breadth[i] = scanner.nextInt();
            height[i] = scanner.nextInt();
        }
        StackTheBoxes result = new StackTheBoxes();
        int answer = result.stackTheBoxes(length,breadth,height,n);
        System.out.println(answer);
        scanner.close();
    }
}

File: Crio/crio_programming_interview_problems-master/StackTheBoxes/StackTheBoxes.cpp
#include<bits/stdc++.h>
using namespace std;
//CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
long long n;
pair<long long,pair<long long,long long> > ar[1009];

bool compare(pair<long long ,pair<long long ,long long > > &i, pair<long long,pair<long long,long long > > &j)
{
    if(i.first>j.first)
        return true;
    else if(i.first==j.first)
    {
        if(i.second.first>j.second.first)
            return true;
        else if(i.second.first==j.second.first)
        {
            if(i.second.second>j.second.second)
                return true;
            else
                return false;
        }
        else
            return false;
    }
    else
        return false;

}
//CRIO_SOLUTION_END_MODULE_L1_PROBLEMS
long long stackTheBoxes(int length[],int breadth[],int height[],int m)
{
    //CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
    for(int i=0;i<m;i++)
    {
        pair<long long,pair<long long,long long> > p;
        p.first=length[i];
        (p.second).first=breadth[i];
        (p.second).second=height[i];
        ar[i+1]=p;
    }
    n=m;
    sort(ar+1,ar+1+n,compare);
    long long i,j,ans=0;
    vector<long long> dp(n+9, 0);
    //long long dp[n+9]={0};
    for(i=1;i<=n;i++)
        dp[i]=ar[i].second.second;
    for(i=n;i>=1;i--)
    {
        for(j=i+1;j<=n;j++)
        {
            if( (ar[j].first < ar[i].first) && ar[j].second.first < ar[i].second.first && ar[j].second.second < ar[i].second.second )
                dp[i]=max(dp[i],dp[j]+ar[i].second.second);
        }
    }
    for(i=1;i<=n;i++)
    {
        ans=max(dp[i],ans);
    }
    return ans;
    //CRIO_SOLUTION_END_MODULE_L1_PROBLEMS
}
int main()
{
    int m;
    cin>>m;
    int length[m],breadth[m],height[m];
    for(long long i=0;i<m;i++)
    {
        cin>>length[i]>>breadth[i]>>height[i];
    }
    cout<<stackTheBoxes(length,breadth,height,m)<<"\n";
}
